pyrival/graphs/maximum_matching.py

```diff
@@ -0,0 +1,65 @@
+import itertools
+from random import randint
+
+
+_DEFAULT_PRIME = 1073750017
+
+
+def maximum_matching(edges, mod=_DEFAULT_PRIME):
+    """
+    Returns the maximum cardinality matching of an undirected general graph
+    Uses a randomized algorithm to compute the rank of the Tutte matrix
+    The rank of the Tutte matrix is equal to twice the size of the maximum matching with high probability
+
+    Complexity: O(n ^ 3) worst case, O(n * m) with high constant factor on average
+
+    :param edges: a list of edges, assume nodes can be anything numbered from 0 to max number in edges
+    :param mod: a large random prime
+    :return: the maximum cardinality matching of the graph
+    """
+
+    edges = [(u, v) for u, v in edges if u != v]
+
+    if not edges:
+        return 0
+
+    n = max(itertools.chain(*edges)) + 1
```

Comment:
Agreed about not making it handle extreme cases, that sounds fair. It will keep the code simpler too. Not so sure about adding `n` as parameter, then we should elaborate whether the graph uses 0 based or 1 based indexing scheme. Is that necessary considering it can be derived easily from the edges?

Reply:
In general pyrival uses 0 based indexing. I definitely think it is fine to require 0 indexing.

Another solution could be to make input be on the same "graph" format (where graph[node] is a list of all its neighbours) as in other graph algorithms in pyrival, for example here https://github.com/cheran-senthil/PyRival/blob/master/pyrival/graphs/lca.py.

Reply:
Not convinced. I understand there are alternative solutions and we can even force 0 based indexing by following pyrival standard. But why would we want to do that? What are the advantages of adding an extra redundant parameter or storing the edges in a less intuitive way here, as opposed to deriving `n` internally with a one-liner from the edges? Even if it's fine/acceptable what you are suggesting, this seems like the better option.